### PR TITLE
REmove complex diagnostic options and options changing reentrancy

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
val build: https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_build/results?buildId=6880004&view=results

### CODE REVIEW

File: src/VisualStudio/LiveShare/Impl/Client/RemoteLanguageServiceWorkspace.cs [hunk 1]
```diff
@@ -537,16 +534,5 @@ private static void UpdateText(ITextBuffer textBuffer, SourceText text)
                 edit.Apply();
             }
         }
-
-        private void StartSolutionCrawler()
-        {
-            if (GlobalOptions.GetOption(SolutionCrawlerRegistrationService.EnableSolutionCrawler))
-            {
-                DiagnosticProvider.Enable(this, DiagnosticProvider.Options.Syntax);
```
Comment (jasonmalinowski) [inline src/VisualStudio/LiveShare/Impl/Client/RemoteLanguageServiceWorkspace.cs:545]: This was only enabling Syntax -- does this need a similar filter for this workspace in that other code?

File: src/Workspaces/Core/Portable/Diagnostics/DiagnosticProvider.cs [hunk 2]
```diff
@@ -1,60 +0,0 @@
...
-using Microsoft.CodeAnalysis.SolutionCrawler;
-
-namespace Microsoft.CodeAnalysis
-{
-    /// <summary>
-    /// Provide a way for users to turn on and off analyzing workspace for compiler diagnostics
-    /// </summary>
-    internal static class DiagnosticProvider
-    {
-        public static void Enable(Workspace workspace, Options options)
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Diagnostics/DiagnosticProvider.cs:18]: Is this what VS Code or VS for Mac are using to enable diagnostics? My worry is we might have to restore this method and take the options but just ignore the options, i.e. delete lines 22-23.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Diagnostics/DiagnosticProvider.cs:18]: i don't see this in Index at all.  @davidwengier do you se any usages of this in vs4mac?

File: src/VisualStudio/LiveShare/Impl/Client/RemoteLanguageServiceWorkspace.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/LiveShare/Impl/Client/RemoteLanguageServiceWorkspace.cs:545]: yes.  i can do that.

Comment (CyrusNajmabadi) [reply src/VisualStudio/LiveShare/Impl/Client/RemoteLanguageServiceWorkspace.cs:545]: done.

**Review by jasonmalinowski — APPROVED** [review]
Approved, assuming we verify no callers to that DiagnosticProvider.Enable that we still need to deal with.

File: src/Workspaces/Core/Portable/Diagnostics/DiagnosticProvider.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Diagnostics/DiagnosticProvider.cs:18]: went with your approach @jasonmalinowski

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/DefaultDiagnosticAnalyzerService.cs [hunk 3]
```diff
@@ -75,23 +75,10 @@ public DefaultDiagnosticIncrementalAnalyzer(DefaultDiagnosticAnalyzerService ser
...
-            }
-
-            private void OnGlobalOptionChanged(object sender, OptionChangedEventArgs e)
-            {
-                if (e.Option == InternalRuntimeDiagnosticOptions.Syntax ||
-                    e.Option == InternalRuntimeDiagnosticOptions.Semantic ||
-                    e.Option == InternalRuntimeDiagnosticOptions.ScriptSemantic)
-                {
-                    var service = _workspace.Services.GetService<ISolutionCrawlerService>();
-                    service?.Reanalyze(_workspace, this, projectIds: null, documentIds: null, highPriority: false);
```
Comment (tmat) [inline src/Features/LanguageServer/Protocol/Features/Diagnostics/DefaultDiagnosticAnalyzerService.cs:93]: >Reanalyze [](http://example.com/codeflow?start=29&length=9)

Shouldn't this be invoked somehow?

Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Features/Diagnostics/DefaultDiagnosticAnalyzerService.cs:93]: we're never changing these options, so unclear why we'd need to even reanalyze.

File: src/Workspaces/Core/Portable/Diagnostics/DiagnosticProvider.cs [hunk 4]
```diff
@@ -1,60 +0,0 @@
...
-        {
-            var service = workspace.Services.GetService<ISolutionCrawlerRegistrationService>();
-
-            var newOptions = GetOptions(workspace, options);
-            workspace.TryApplyChanges(workspace.CurrentSolution.WithOptions(newOptions));
-            service.Register(workspace);
```
Comment (tmat) [inline src/Workspaces/Core/Portable/Diagnostics/DiagnosticProvider.cs:24]: >service.Register(workspace) [](http://example.com/codeflow?start=12&length=27)

service.Register should probably still be called

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Diagnostics/DiagnosticProvider.cs:24]: done.

**Review by jasonmalinowski — APPROVED** [review]

**Review by tmat — APPROVED** [review]
:shipit:

File: src/VisualStudio/Core/Test/Diagnostics/DefaultDiagnosticUpdateSourceTests.vb [hunk 5]
```diff
@@ -92,8 +90,6 @@ class A
                 Dim miscService = GetDefaultDiagnosticAnalyzerService(workspace)
                 Assert.False(miscService.SupportGetDiagnostics)
 
-                DiagnosticProvider.Enable(workspace, DiagnosticProvider.Options.Syntax)
```
Comment (tmat) [inline src/VisualStudio/Core/Test/Diagnostics/DefaultDiagnosticUpdateSourceTests.vb:95]: > DiagnosticProvider.Enable(workspace, [](http://example.com/codeflow?start=15&length=37)

probably want to keep the call w/o options

File: src/VisualStudio/Core/Test/Diagnostics/DefaultDiagnosticUpdateSourceTests.vb [hunk 6]
```diff
@@ -131,8 +127,6 @@ class A
                 Dim miscService = GetDefaultDiagnosticAnalyzerService(workspace)
                 Assert.False(miscService.SupportGetDiagnostics)
 
-                DiagnosticProvider.Enable(workspace, DiagnosticProvider.Options.Semantic)
```
Comment (tmat) [inline src/VisualStudio/Core/Test/Diagnostics/DefaultDiagnosticUpdateSourceTests.vb:134]: >Enable [](http://example.com/codeflow?start=35&length=6)

probably want to keep the call w/o options

**Comment by CyrusNajmabadi** [comment]
Val build: https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_build/results?buildId=6880004&view=logs&s=d00c89cd-5e0a-588d-122b-bf62b983d932

**Comment by CyrusNajmabadi** [comment]
Insertion: https://dev.azure.com/devdiv/DevDiv/_git/VS/pullrequest/432045

**Comment by CyrusNajmabadi** [comment]
All internal tests passing.  Merging in.

**Pull request merged**